bigquery_etl/dryrun.py

```diff
@@ -225,6 +243,145 @@ def get_sql(self):
+            if file_age > ttl_seconds:
+                try:
+                    cache_file.unlink()
+                except OSError:
+                    pass
+                return None
+
+            cached_data = pickle.loads(cache_file.read_bytes())
+            cache_age = time.time() - cache_file.stat().st_mtime
+            print(f"[DRYRUN CACHE HIT] {self.sqlfile} (age: {cache_age:.0f}s)")
+            return cached_data
+        except (pickle.PickleError, EOFError, OSError, FileNotFoundError) as e:
+            print(f"[DRYRUN CACHE] Failed to load cache: {e}")
+            try:
+                if cache_file.exists():
+                    cache_file.unlink()
+            except OSError:
+                pass
+            return None
+
+    def _save_cached_result(self, cache_key, result):
+        """Save dry run result to disk cache using atomic write."""
+        cache_dir = Path(tempfile.gettempdir()) / "bigquery_etl_dryrun_cache"
+        cache_dir.mkdir(parents=True, exist_ok=True)
+        cache_file = cache_dir / f"dryrun_{cache_key}.pkl"
+
+        try:
+            # write to temporary file first, then atomically rename
+            # this prevents race conditions where readers get partial files
+            temp_file = Path(str(cache_file) + f".tmp.{os.getpid()}")
+            with open(temp_file, "wb") as f:
+                pickle.dump(result, f)
+                f.flush()
+                os.fsync(f.fileno())  # Ensure data is written to disk
+
+            temp_file.replace(cache_file)
+
+            # save table metadata separately if present
+            if (
+                result
+                and "tableMetadata" in result
+                and self.project
+                and self.dataset
+                and self.table
+            ):
+                table_identifier = f"{self.project}.{self.dataset}.{self.table}"
+                self._save_cached_table_metadata(
+                    table_identifier, result["tableMetadata"]
+                )
+        except (pickle.PickleError, OSError) as e:
+            print(f"[DRYRUN CACHE] Failed to save cache: {e}")
+            try:
+                temp_file = Path(str(cache_file) + f".tmp.{os.getpid()}")
+                if temp_file.exists():
+                    temp_file.unlink()
+            except OSError:
+                pass
+
+    def _get_cached_table_metadata(self, table_identifier, ttl_seconds=None):
+        """Load cached table metadata from disk based on table identifier."""
+        if ttl_seconds is None:
+            ttl_seconds = ConfigLoader.get("dry_run", "cache_ttl_seconds", fallback=900)
+
+        cache_dir = Path(tempfile.gettempdir()) / "bigquery_etl_dryrun_cache"
+        cache_dir.mkdir(parents=True, exist_ok=True)
+        # table identifier as cache key
+        table_cache_key = hashlib.sha256(table_identifier.encode()).hexdigest()
+        cache_file = cache_dir / f"table_metadata_{table_cache_key}.pkl"
+
+        try:
```

Comment:
The cache read and write code is almost identical between the metadata and the dry run result.  Moving the common code into a common function would make this easier to maintain